libs/cardano-ledger-conformance/src/Test/Cardano/Ledger/Conformance/SpecTranslate/Conway.hs

```diff
@@ -0,0 +1,875 @@
+import Cardano.Ledger.Coin (Coin (..))
+import Cardano.Ledger.Conway.Core
+import Cardano.Ledger.Conway.Governance (
+  Committee (Committee),
+  Constitution (..),
+  EnactState (..),
+  GovAction (..),
+  GovActionId (GovActionId),
+  GovActionIx (..),
+  GovActionState (..),
+  GovProcedures (..),
+  GovPurposeId (GovPurposeId),
+  GovRelation (..),
+  ProposalProcedure (..),
+  Proposals,
+  Vote (..),
+  Voter (..),
+  VotingProcedure (..),
+  VotingProcedures (..),
+  foldrVotingProcedures,
+  gasAction,
+  gasReturnAddr,
+  pPropsL,
+ )
+import Cardano.Ledger.Conway.PParams (ConwayPParams (..), THKD (..))
+import Cardano.Ledger.Conway.Rules (ConwayGovPredFailure, GovEnv (..))
+import Cardano.Ledger.Conway.Scripts (ConwayPlutusPurpose (..))
+import Cardano.Ledger.Conway.TxCert (
+  ConwayDelegCert (..),
+  ConwayGovCert (..),
+  ConwayTxCert (..),
+  getStakePoolDelegatee,
+  getVoteDelegatee,
+ )
+import Cardano.Ledger.Credential (Credential (..))
+import Cardano.Ledger.Crypto (Crypto (..))
+import Cardano.Ledger.DRep (DRep (..))
+import Cardano.Ledger.HKD (HKD)
+import Cardano.Ledger.Keys (KeyHash (..), VKey (..))
+import Cardano.Ledger.Keys.WitVKey (WitVKey (..))
+import Cardano.Ledger.Plutus (CostModels, ExUnits (..), Prices)
+import Cardano.Ledger.Plutus.Data (Data, Datum (..))
+import Cardano.Ledger.PoolParams (PoolParams (..))
+import Cardano.Ledger.SafeHash (SafeHash, extractHash)
+import Cardano.Ledger.Shelley.LedgerState (UTxOState (..))
+import Cardano.Ledger.Shelley.Rules (Identity, UtxoEnv (..))
+import Cardano.Ledger.Tools (byteStringToNum)
+import Cardano.Ledger.TxIn (TxId (..), TxIn (..))
+import Cardano.Ledger.UTxO (UTxO (..))
+import Cardano.Ledger.Val (Val (..))
+import Constrained (HasSimpleRep, HasSpec)
+import Control.DeepSeq (NFData)
+import Control.Monad.Except (MonadError (..))
+import Control.State.Transition.Extended (STS (..))
+import Data.Bifunctor (Bifunctor (..))
+import Data.Bitraversable (bimapM)
+import Data.Data (Typeable)
+import Data.Foldable (Foldable (..))
+import qualified Data.List as L
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import Data.OMap.Strict (OMap, assocList)
+import Data.OSet.Strict (OSet)
+import Data.Ratio (denominator, numerator)
+import Data.Sequence.Strict (StrictSeq)
+import Data.Set (Set)
+import qualified Data.Set as Set
+import Data.Traversable (forM)
+import Data.Word (Word32, Word64)
+import GHC.Generics (Generic)
+import Lens.Micro
+import Lens.Micro.Extras (view)
+import qualified Lib as Agda
+import Test.Cardano.Ledger.Conformance (
+  SpecTransM,
+  SpecTranslate (..),
+  SpecTranslationError,
+  askTransCtx,
+ )
+import Test.Cardano.Ledger.Constrained.Conway.Instances (IsConwayUniv)
+import Test.Cardano.Ledger.Conway.TreeDiff (ToExpr (..))
+
+-- | OpaqueErrorString behaves like unit in comparisons, but contains an
+-- error string that can be displayed.
+newtype OpaqueErrorString = OpaqueErrorString String
```

Comment:
The predicate failures are represented as strings in the Agda spec, we might change them to something better in the future. For now all failures are treated as a unit type, so all failures are treated equally, but with this type I can still print out information about which part of the Agda spec failed. Once we

Reply:
> The spec can actually give error messages when any of the predicates are not satisfied

Aha, that is something that I did not realize. My misconception was that Agda just doesn't build if any pf the predicate checks is not satisfied. It would make sense that the Haskell representation of the spec would produce a runtime failure instead for this case.

> Once we have better types for predicate failures in the Agda spec we can get rid of this type.

Sounds like a good plan

Reply:
On background (and to check my understanding and recollection, please correct me if Im wrong @Soupstraw): the way the Agda side works (I paired a but with Ulf on this) is basically that there is a proof that the rule is functional (that is in the functional relation) sense of the word. If you have an environment, a state, and a signal you can uniquely determine either a failure to transition or the goal state. The bit that is certified here is that the function that computes this on the agda side is proven to be sound and complete w.r.t. the rule on the agda side.

This function and the relevant types are then compiled from Agda to Haskell and can be invoked in these tests as a Haskell function that implements the STS transition.

Reply:
@MaximilianAlgehed Yeah, that's how it works. The rules on the Agda side use a relational approach, but every rule has a `Computational` typeclass. That typeclass has a method that determines the output from all the inputs, and a proof that this method returns a value iff the arguments and the output are related via the relation.